### PR TITLE
AP-2832

### PR DESCRIPTION
**Pull request by chiichang**
- Incorrect arc dur val caused by mis-assigned overlapped events to activities.
- Fixed with new unit test added

### CODE REVIEW

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java [hunk 1]
```diff
@@ -252,4 +255,39 @@ public static void testRetain3(APMLog apmLog, APMLogUnitTest parent) throws Unsu
...
+    public static double getTotal(DoubleArrayList doubleList) {
+        if (doubleList.size() > 0) {
+            double sum = 0;
+            for (int i = 0; i < doubleList.size(); i++) {
+                sum += doubleList.get(i);
+            }
+            return sum;
+        }
+        return 0;
+    }
```
Comment (raboczi) [inline Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: I think you may find that DoubleArrayList has a .sum() method which makes this method unnecessary.

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java [hunk 2]
```diff
@@ -252,4 +255,39 @@ public static void testRetain3(APMLog apmLog, APMLogUnitTest parent) throws Unsu
...
+        }
+    }
+
+    public static double getAverage(DoubleArrayList doubleList) {
+        if (doubleList.size() > 0) {
+            double sum = getTotal(doubleList);
+            return doubleList.size() > 1 ? sum / doubleList.size() : sum;
+        }
+        return 0;
+    }
```
Comment (raboczi) [inline Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: I think there's a provided DoubleArrayList.average method to do this without an extra method.

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java [hunk 3]
```diff
@@ -252,4 +255,39 @@ public static void testRetain3(APMLog apmLog, APMLogUnitTest parent) throws Unsu
...
+        DurSubGraph subGraph = pLog.getAttributeGraph().getNextValueDurations("concept:name", "A");
+        DoubleArrayList durList = subGraph.getValDurListMap().get("A");
+        double avg = getAverage(durList);
+        double twoHours = 1000 * 60 * 60 * 2d;
+        if (avg != twoHours) {
+            throw new AssertionError("TEST FAILED. RESULT MISMATCH.\n expected value=" + twoHours +
+                    "; output value=" + avg);
+        } else {
+            parent.printString("'Attribute arc duration (4) - Average duration' test PASS.\n");
+        }
```
Comment (raboczi) [inline Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: JUnit has various assertion methods which could simplify test code like this, something like:

```
import static org.junit.Assert.assertEquals;
```
```
assertEquals(1000 * 60 * 60 * 2d, getAverage(durList));
```

**Review by raboczi — COMMENTED** [review]
Can we arrange a quick walkthrough of this code so that I'm clear about the problem it addresses?  I'm guessing it's about the duration of overlapping events but it doesn't actually mention in the test comments, etc.  This is pretty short so it ought to be quick.

**Comment by brucenguyen1** [comment]
> Can we arrange a quick walkthrough of this code so that I'm clear about the problem it addresses? I'm guessing it's about the duration of overlapping events but it doesn't actually mention in the test comments, etc. This is pretty short so it ought to be quick.

Yes, I tried to understand the changes to do a review but had no clues. The changes in this PR are only a few but the existing code is much bigger, a quick walkthrough to share how things are organized and related here would be helpful to understand why these changes were made to fix the issue, and so why the unit tests were written to cover this scenario.

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java (hunk 3 above)
Comment (chiichang) [reply Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: I'm not sure where you want to place that code. Please edit it directly.

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java (hunk 2 above)
Comment (chiichang) [reply Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: updated

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java (hunk 1 above)
Comment (chiichang) [reply Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: removed

**Comment by chiichang** [comment]
> > Can we arrange a quick walkthrough of this code so that I'm clear about the problem it addresses? I'm guessing it's about the duration of overlapping events but it doesn't actually mention in the test comments, etc. This is pretty short so it ought to be quick.
> 
> Yes, I tried to understand the changes to do a review but had no clues. The changes in this PR are only a few but the existing code is much bigger, a quick walkthrough to share how things are organized and related here would be helpful to understand why these changes were made to fix the issue, and so why the unit tests were written to cover this scenario.

The algorithm starts counting the events not from "start" but from any lifecycle:transition value, which means in the old code, to get the events of an activity, it took all the events that have same concept:name  (including "start") after the first event until it reaches a "complete" or the end of the case. 

That causes problem and is explicitly shown in the arc duration filter (the values derived from DurSubGraph). Hence, the unit test is just to test the value from the DurSubGraph.

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java (hunk 3 above)
Comment (raboczi) [reply Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java]: Done!

**Review by raboczi — APPROVED** [review]
I'm satisfied that my suggestions have been addressed, and that this builds, passes its tests, and runs.

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java [hunk 4]
```diff
@@ -415,7 +417,14 @@ private AActivity getActivity(int index, int fromIndex, IntArrayList markedIndex
...
+                                if (!foundStart) {
+                                    eventIndexList.add(i);
+                                    foundStart = true;
+                                }
+                            } else {
+                                eventIndexList.add(i);
+                            }
                         }
                     }
                 }
```
Comment (brucenguyen1) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java:430]: This code is duplicated with the code above in LogFactory.java (getActivity). Can we do a bit refactoring here to remove this duplication?

**Pull request merged**